pymodbus/client/base.py

```diff
@@ -171,10 +171,16 @@ async def async_execute(self, request) -> ModbusResponse:
             except asyncio.exceptions.TimeoutError:
                 count += 1
         if count > self.retries:
-            self.close(reconnect=True)
-            raise ModbusIOException(
-                f"ERROR: No response received after {self.retries} retries"
-            )
+            if request.timeout_handler is not None:
+                if asyncio.iscoroutinefunction(request.timeout_handler):
```

Comment:
> callbacks start with "on_",

Changed to "on_".

> please simply decide on one type or the other.

Since this is an application defined callback I could see both sync or async callbacks being needed to successfully handle a timeout.